relay-server/src/actors/upstream.rs

```diff
@@ -881,69 +890,115 @@ impl Entry {
+                    RequestPriority::Low => self.low.push_back(entry),
+                };
+            }
+            RequestAttempt::Retry => {
+                match priority {
+                    RequestPriority::High => self.retries.push_front(entry),
+                    RequestPriority::Low => self.retries.push_back(entry),
+                };
+            }
         }
 
         relay_statsd::metric!(
             histogram(RelayHistograms::UpstreamMessageQueueSize) = self.len() as u64,
             priority = priority.name(),
+            attempt = attempt.name()
         );
     }
 
     /// Places an entry at the back of the queue.
     ///
-    /// Since entries are dequeued in FIFO order, this entry will be dequeued last within its
-    /// priority class.
+    /// Since entries are dequeued in FIFO order, this entry will be dequeued
+    /// last within its priority class; see
+    /// [`dequeue`][`UpstreamQueue::dequeue`] for more details.
     pub fn enqueue(&mut self, entry: Entry) {
-        self.put(entry, EnqueuePosition::Back)
+        self.put(entry, RequestAttempt::First)
     }
 
-    /// Place an entry at the front of the queue.
+    /// Places an entry in the retry queue.
     ///
-    /// This entry will be dequeued next within its priority class, unless another call to
-    /// `enqueue_immediate` follows.
-    pub fn enqueue_immediate(&mut self, entry: Entry) {
-        self.put(entry, EnqueuePosition::Front)
+    /// If the entry is high priority it's placed at the front of the queue, or
+    /// at the back if it's low priority. Entries in the retry queue are
+    /// dequeued from front to back.
+    ///
+    /// It also schedules the next retry time, based on the retry back off. The
+    /// retry queue is not dequeued until the next retry has elapsed.
+    pub fn enqueue_retry(&mut self, entry: Entry) {
+        self.put(entry, RequestAttempt::Retry);
+        self.next_retry = Instant::now() + self.retry_backoff.next_backoff();
     }
 
     /// Removes the head of the queue with highest priority.
     ///
-    /// This always returns entries with [high](RequestPriority::High) first before dequeueing
-    /// entries with low priority.
+    /// Dequeues entries in the following order:
+    /// 1. [High priority][RequestPriority::High] requests.
+    /// 2. [High priority][RequestPriority::High]
+    /// [retries][RequestAttempt::Retry], after backoff elapsed.
+    /// 3. [Low priority][RequestPriority::High]
+    /// [retries][RequestAttempt::Retry], after backoff elapsed.
+    /// 4. [Low priority][RequestPriority::Low] requests.
     pub fn dequeue(&mut self) -> Option<Entry> {
-        self.high.pop_front().or_else(|| self.low.pop_front())
+        if !self.high.is_empty() {
+            return self.high.pop_front();
+        }
+        if !self.retries.is_empty() && self.next_retry <= Instant::now() {
+            return self.retries.pop_front();
+        }
+        self.low.pop_front()
```

Comment:
This should _first_ check the retries (but with separate queues for high and low), and then the new queues, otherwise incoming requests can starve retries. Essentially your logic should be:

```rust
let mut entry = None;

if self.next_retry <= Instant::now() {
  entry = self.retry_high.pop_front().or_else(|| self.retry_low.pop_front());
}

entry.or_else(|| self.high.pop_front()).or_else(|| self.low.pop_front())
```

Reply:
This is a trade-off and really depends on what we want to prioritize. Let's consider the following scenarios, assuming Relay produces more low-priority requests than high-priority (e.g. more outgoing envelopes than project config requests):

1. Under regular operation, everything works as expected.
2. When upstream is completely down, the order doesn't matter.
3. When upstream is flaky but downstream works fine, Relay is still processing envelopes but needs to decide what is prioritizing. If high-priority requests are prioritized, there's a chance that envelopes starve; if low-priority requests are prioritized, there's a chance we don't get project config updates (or even new ones).
4. When upstream recovers after being down, all requests will eventually go out and we need to decide what to request first. Requesting high-priority first means prioritizing config fetching and maybe more accurate envelope processing; requesting low-priority first means flushing envelopes but possibly processing incoming envelopes with stale configs.

Another possibility is the initial assumption to be incorrect and Relay produces more high-priority requests than low-priority ones. That is a different issue that should be handled independently.

In my opinion, keeping envelopes in memory and ensuring up-to-date envelope processing is more important. I'm optimistic and think envelopes will be forwarded in the end, but if this can really be an issue we may want to consider spooling them at this stage. I've followed this approach in [36c065a](https://github.com/getsentry/relay/pull/2518/commits/36c065a7a3b171723981c8421bad7fcf0ce1410e). I'm not strongly opinionated and happy to change the prioritization if somebody thinks otherwise or I'm missing something.